express_relay/programs/express_relay/src/lib.rs

```diff
@@ -0,0 +1,352 @@
+use anchor_lang::solana_program::sysvar::instructions as sysvar_instructions;
+use solana_program::{serialize_utils::read_u16, sysvar::instructions::{get_instruction_relative, load_instruction_at_checked}};
+use anchor_syn::codegen::program::common::sighash;
+use anchor_spl::token::Token;
+use crate::{
+    error::ErrorCode,
+    state::*,
+    utils::*,
+};
+
+declare_id!("GwEtasTAxdS9neVE4GPUpcwR7DB7AizntQSPcG36ubZM");
+
+#[program]
+pub mod express_relay {
+    use super::*;
+
+    pub fn initialize(ctx: Context<Initialize>, data: InitializeArgs) -> Result<()> {
+        validate_fee_split(data.split_protocol_default)?;
+        validate_fee_split(data.split_relayer)?;
+
+        let express_relay_metadata_data = &mut ctx.accounts.express_relay_metadata;
+        express_relay_metadata_data.admin = *ctx.accounts.admin.key;
+        express_relay_metadata_data.relayer_signer = *ctx.accounts.relayer_signer.key;
+        express_relay_metadata_data.fee_receiver_relayer = *ctx.accounts.fee_receiver_relayer.key;
+        express_relay_metadata_data.split_protocol_default = data.split_protocol_default;
+        express_relay_metadata_data.split_relayer = data.split_relayer;
+
+        Ok(())
+    }
+
+    pub fn set_admin(ctx: Context<SetAdmin>) -> Result<()> {
+        let express_relay_metadata_data = &mut ctx.accounts.express_relay_metadata;
+
+        express_relay_metadata_data.admin = *ctx.accounts.admin_new.key;
+        Ok(())
+    }
+
+    pub fn set_relayer(ctx: Context<SetRelayer>) -> Result<()> {
+        let express_relay_metadata_data = &mut ctx.accounts.express_relay_metadata;
+
+        express_relay_metadata_data.relayer_signer = *ctx.accounts.relayer_signer.key;
+        express_relay_metadata_data.fee_receiver_relayer = *ctx.accounts.fee_receiver_relayer.key;
+
+        Ok(())
+    }
+
+    pub fn set_splits(ctx: Context<SetSplits>, data: SetSplitsArgs) -> Result<()> {
+        validate_fee_split(data.split_protocol_default)?;
+        validate_fee_split(data.split_relayer)?;
+
+        let express_relay_metadata_data = &mut ctx.accounts.express_relay_metadata;
+
+        express_relay_metadata_data.split_protocol_default = data.split_protocol_default;
+        express_relay_metadata_data.split_relayer = data.split_relayer;
+
+        Ok(())
+    }
+
+    pub fn set_protocol_split(ctx: Context<SetProtocolSplit>, data: SetProtocolSplitArgs) -> Result<()> {
+        validate_fee_split(data.split_protocol)?;
+
+        ctx.accounts.protocol_config.protocol = *ctx.accounts.protocol.key;
+        ctx.accounts.protocol_config.split = data.split_protocol;
+
+        Ok(())
+    }
+
+    pub fn permission(ctx: Context<Permission>, data: PermissionArgs) -> Result<()> {
+        if data.deadline < Clock::get()?.unix_timestamp {
+            return err!(ErrorCode::DeadlinePassed);
+        }
+
+        // check that not cpi
+        let instruction = get_instruction_relative(0, &ctx.accounts.sysvar_instructions.to_account_info())?;
```

Comment:
I just found this:
https://docs.rs/solana-program/latest/solana_program/instruction/fn.get_stack_height.html

Seems like an easier way to detect cpi by doing something like this:
```
if get_stack_height() != TRANSACTION_LEVEL_STACK_HEIGHT {
      return err!(ErrorCode::InvalidCPIPermission);
  }
```